maildir.go

```diff
@@ -502,6 +500,44 @@ func (d Dir) Clean() error {
 	return nil
 }
 
+// Walk calls fn for every messages in the mailbox's subdir directory.
+//
+// subdir should be "cur", "new" or "tmp".
+//
+// It stops on errors (parsing filename, flags and fn returning an error).
+func (d Dir) Walk(subdir string, fn func(key string, flags []Flag) error) error {
+	f, err := os.Open(filepath.Join(string(d), subdir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for {
+		names, err := f.Readdirnames(readdirChunk)
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		for _, filename := range names {
+			key, err := parseKey(filename)
+			if err != nil {
+				return err
+			}
+			flags, err := d.Flags(key)
```

Comment:
We should be able to get the flags from the filename directly, without needing to go back to guessing the filename from the key in `Dir.Flags`. I'd suggest extracting some of the `Dir.Flags` logic into a `parseFlags(filename string) ([]Flag, error)` function.